### ISSUE TITLE
5.0 RC1 System.NullReferenceException on Add Migration

### ISSUE THREAD

**Issue by jaysee260** [dotnet/efcore#22619]
I upgraded from EF Core 3.1 to 5.0 RC 1 to test out the new way it handles many-to-many relationship.
When I try adding my first migration, I am getting a System.NullReferenceException.

While on 3.1, I had a few more models that acted as join tables between the various entities... `BookAuthorEntity`, `BookPublisherEntity`, `BookKeywordEntity`, etc. Migrations worked fine.

After upgrading to 5.0 RC 1, I got rid of them and followed the setup suggested by the [EF Core 5.0 RC1 announcement post](https://devblogs.microsoft.com/dotnet/announcing-entity-framework-core-efcore-5-0-rc1/#many-to-many) on how to handle Many-to-Many.

Here's some of my code to help with context:

<details>
  <summary>Models</summary>

  ```csharp
    [Table("Author")]
    public class AuthorEntity
    {
        [Key]
        public Guid AuthorId { get; set; }
        [Required]
        public string FirstName { get; set; }
        [Required]
        public string LastName { get; set; }

        // Relationships
        public ICollection<BookEntity> Books { get; set; }
    }

    [Table("Translator")]
    public class TranslatorEntity
    {
        [Key]
        public Guid TranslatorId { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }

        // Relationships
        public ICollection<BookEntity> Books { get; set; }
    }

    [Table("Editor")]
    public class EditorEntity
    {
        [Key]
        public Guid EditorId { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }

        // Relationships
        public ICollection<BookEntity> Books { get; set; }
    }

    [Table("Publisher")]
    public class PublisherEntity
    {
        [Key]
        public Guid PublisherId { get; set; }
        [Required]
        public string Name { get; set; }

        // Relationships
        public ICollection<BookEntity> Books { get; set; }
    }

    [Table("Keyword")]
    public class KeywordEntity
    {
        [Key]
        public Guid KeywordId { get; set; }
        [Required]
        public string Keyword { get; set; }

        // Relationships
        public ICollection<BookEntity> Books { get; set; }
    }

    [Table("Book")]
    public class BookEntity
    {
        [Key]
        public Guid BookId { get; set; }
        [Required]
        public string Title { get; set; }
        public string? Subtitle { get; set; }
        public string ISBN { get; set; }
        public int? PublicationYear { get; set; }
        public int? TranslationPublicationYear { get; set; }

        // Relationships
        public ICollection<AuthorEntity> Authors { get; set; }
        public ICollection<TranslatorEntity> Translators { get; set; }
        public ICollection<EditorEntity> Editors { get; set; }
        public PublisherEntity Publisher { get; set; }
        public ICollection<KeywordEntity> Keywords { get; set; }
    }

  ```
</details>

<details>
  <summary>DbContext</summary>

  ```csharp
    public class LibraryContext : DbContext
    {
        // Entity tables
        public DbSet<BookEntity> Books { get; set; }
        public DbSet<AuthorEntity> Authors { get; set; }
        public DbSet<TranslatorEntity> Translators { get; set; }
        public DbSet<EditorEntity> Editors { get; set; }
        public DbSet<PublisherEntity> Publishers { get; set; }
        public DbSet<KeywordEntity> Keywords { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseNpgsql("Host=localhost;Port=5432;Database=test;Username=test;Password=test");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.HasPostgresExtension("uuid-ossp");
        }

  ```
</details>

<details>
  <summary>Program</summary>

  ```csharp
    class Program
    {
        static void Main(string[] args) {}
    }
  ```
</details>

### Steps to reproduce

After running `dotnet ef migrations add InitializeDatabase`, I get the following output:

```
Build started...
Build succeeded.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.ForeignKeyPropertyDiscoveryConvention.DiscoverProperties(IConventionForeignKeyBuilder relationshipBuilder, IConventionContext context)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.ForeignKeyPropertyDiscoveryConvention.ProcessSkipNavigationForeignKeyChanged(IConventionSkipNavigationBuilder skipNavigationBuilder, IConventionForeignKey foreignKey, IConventionForeignKey oldForeignKey, IConventionContext`1 context)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnSkipNavigationForeignKeyChanged(IConventionSkipNavigationBuilder navigationBuilder, IConventionForeignKey foreignKey, IConventionForeignKey oldForeignKey)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnSkipNavigationForeignKeyChangedNode.Run(ConventionDispatcher dispatcher)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.DelayedConventionScope.Run(ConventionDispatcher dispatcher)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Run()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Dispose()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnModelInitialized(IConventionModelBuilder modelBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnModelInitialized(IConventionModelBuilder modelBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.Model..ctor(ConventionSet conventions, ModelDependencies modelDependencies)
   at Microsoft.EntityFrameworkCore.ModelBuilder..ctor(ConventionSet conventions, ModelDependencies modelDependencies, Boolean _)
   at Microsoft.EntityFrameworkCore.ModelBuilder..ctor(ConventionSet conventions, ModelDependencies modelDependencies)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, ModelDependencies modelDependencies)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder, ModelDependencies modelDependencies)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_3(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Object reference not set to an instance of an object.
```

### Further technical details

EF Core version: 5.0.0-rc.1.20451.13
Database provider: Npgsql.EntityFrameworkCore.PostgreSQL - 5.0.0-rc1
Target framework: .NET Core 3.1
Operating system: Windows 10 - Version 2004, OS Build 19041.450
IDE: Visual Studio Code 1.48.2

**Comment by smitpatel** [comment]
Possible duplicate of #22494

**Comment by ajcvickers** [comment]
Verified this is fixed with the latest [daily build](https://github.com/dotnet/efcore/blob/main/docs/DailyBuilds.md).

**Comment by programmer1718** [comment]
Thank you for the prompt replies @smitpatel and @ajcvickers.
Using the latest daily build, I don't have this problem anymore.

Any clue of when this fix might make it into a Nuget package?

**Comment by ajcvickers** [comment]
@programmer1718 It will be included in 5.0 RC2. (Following the link to the duplicate issue shows that it is marked as `closed-fixed` and the milestone is `5.0.0-rc2`.)

**Comment by pantonis** [comment]
I can verify that this was fixed in latest daily build
